### ISSUE TITLE
RFC: Lesson Planning

### ISSUE THREAD

**Issue by chrsep** [obserfy/obserfy#385]
Lesson Plans are a list of materials/lessons/activity that teachers/parent can do for a given day. It could repeat in a weekly pattern and usually have some accompanying materials (pdfs, instructions, etc).

 It is prepared in advanced so that when the day comes, they could just pick something from the list of lessons/material. It could have relations to some observations and some curriculum materials. Though for now, we should try to build the simplest form of this feature.

Important parts of this feature:
1. Overview page, teachers need to be able to see an overview of plans for a certain date.
1. Targeting, to whom the plans are going to be applied to
2. Repetition, able to automatically repeat.
3. File attachments

### Overview

Teachers need to be able to see all of the plans that have been set for a certain date so that they can act accordingly.

### Targeting

Lesson plans are about what the teachers going to do with a certain student on a certain day. Because of this, teachers will need to be able to set which students are being targeted by a plan.

To make this easy, we should **allow teachers to target multiple students**, either manually or with certain criteria, such as class.

This is also important because **parents would need to be able to see all of the plans** that are being targeted at their child so they can have a clear view of their child's activity.

### Repetition

A plan might need to be done daily or weekly or monthly for a certain period of time. Every instance of a repeated plan needs to have consistent details (same title, desc, files, etc...).

This means a plan should be able to **accept a type of repetition** (0: daily, 1: weekly: 2: monthly), **and end date**, a date when the repetition is going to stop. (The point is, a plan needs to be able to repeat for a certain amount of time in a certain pattern, there might be better ways to do this).

### File Attachments

A plan might have some files attached to it. Teachers should be able to upload their own files or select some from a list of files that have already been uploaded before.

A single file can have none, 1, or multiple files attached with the following extensions:
- pdfs
- jpg/png/SVG

## Execution

We should prioritize and do this feature in stages so we could ship this early. Following is a suggestion on which things to be prioritized first.

### Priority 1
- [ ] Overview page. [storybook](https://sb.obserfy.com/?path=/story/page-plans-pageplans--with-layout)
- [ ] Create new plan Page [storybook](https://sb.obserfy.com/?path=/story/page-plans-pagenewplan--with-layout)
- [ ] Plan details page [storybook](https://sb.obserfy.com/?path=/story/page-plans-pagplandetails--with-layout)
- [ ] One-off plans (happens only once, on a certain date)
- [ ] Able to set Title
- [ ] Able to set description
- [ ] Able to target a class
- [ ] Able to attach and upload a single file

#### Related endpoints
`DELETE` `/plans/{planId}`
`PATCH` `/plans/{planId}`
`POST` `/classes/{classId}/plans`
`GET` `/schools/{schoolId}/plans?date=2020-20-02`

### Priority 2
- [ ] Able to do weekly repetition
- [ ] Able to target individual students
- [ ] Able to attach multiple files
- [ ] Able to attach past uploaded files

#### Related endpoints
`GET` `/schools/{schoolId}/files` list of uploaded files for attachments
`POST` `/schools/{schoolId}/files` upload a file

### Wireframe:
![image](https://user-images.githubusercontent.com/8491552/81610462-c6d6cc80-9403-11ea-821d-877197534d2e.png)

**Comment by chrsep** [comment]
I've been thinking about how to implement a repeating lesson and I need some feedback. Here's what I've come up with. There might still be a better/simpler way than this though.

#### Solution 1. Saved in a simple column
We could simply add a column (repetition, and end date) on the lesson plan's table. Then, when the client calls `get plans`, we'll construct each of the plan's instances in memory (for example. if repeat every Monday, API will return `[{data for Monday 18th May}, {data for Monday 25th May}, {data for Monday 1st}, ...]`

But this has some drawbacks:
1. Expensive `get` requests because we'll need to reconstruct the instances every call.
2. It's complicated to store data related to each instance (such as assignee, related observations, etc...) in the future. Since the instances can't be referenced directly in DB.
3. Can't delete any of the instances since it is generated on runtime. (well.... we could delete it, by using some sort of a whitelist table, but this would just be madness)

#### Solution 2. Save every instance in DB

To solve solution 1's drawback, we could instead save the instances on the DB directly. So on `POST` new plan call, we'll generate every instance of the lesson plan (the  `[{data for Monday 18th May}, {data for Monday 25th May}, {data for Monday 1st}, ...]`  and save it to DB.

This way, every instance can have their own metadata down the road (related observations, assignees, done/not done, etc...) and be able to be deleted individually. But it still has its own drawbacks:

1. Repeated lesson plans would lose their relationship with one another.
2. We'll lose the repetition pattern of every lesson plan.
3. Mistakes are harder to fix. Users would need to delete/edit every instance manually when they added a wrong plan.

#### Solution 3. Save every instance in DB but separate out the lesson details

To solve solution 2's problem, we could separate out the lesson plan details (title, descriptions, attachments, repetition pattern, etc...) into its own table, and reference its key on each of its instances. 

This would retain the relationship between related lesson plans. No duplicated data makes editing easier. And since we know which lesson plan instances are related to one another, we could add a more sophisticated deleting mechanism (delete every instance after this, delete all instance of this plan, etc...).

I think solution 3 might be the best way to go for now. But it will probably a little bit more complex to implement upfront. But still, I'm happy to hear any simpler approach.

**Comment by chrsep** [comment]
Here is more detail on file uploads. 

### Priorities
1. Lesson Plan can have 1 or more file attached to it
2. One file can be attached to multiple lesson plans
3. Unlike images, the filename is important because this is how our user is going to identify the file
4. User can choose to attach files uploaded in the past or upload a new one to a lesson plan (optional, we could do this later)
5. User can view every file uploaded to the school (required by no. 4)

#### Solution 1
We could implement this in a similar way with how we currently handle images. Upload to S3 with `/schooldId/file_name`, then save only the full key to DB in something like fileToLessonPlan.

##### Pros
- Very simple

##### Cons
- Renaming file will require us to modify s3 file, then update the database column
- We'll need to check to s3 if a file with the same name already exists every time users upload a new file to prevent overwriting pre-existing files.

#### Solution 2
We could save it to S3 with `/schoolId/fileId` as the key to s3, with fileId being the ID of the file on DB. We'll then save the filename inside the DB too. We could mark the filename as unique to prevent file management chaos since duplicated names can be confusing. 

When the user tries to download the file, we'll use S3's content-disposition query string to set the file name.

So here we'll have 3 tables:
- Files (with id, schoolId, and filename as columns)
- FilesToLessonPlan (for many to many relationship)
- LessonPlans

##### Pros
- Editing name is simple, one update to file table and the file is updated.
- Since filename is a unique column, we'll figure out if user is uploading the same file twice directly when we try to insert to the DB
- It'll be easy to add metadata to a file, such as related tags, etc...

##### Cons
- Don't see any yet hmm
